### PR TITLE
fix(sent sms list): store filter and page params in URL (DHIS2-9866)

### PR DESCRIPTION
**Pull request by mediremi**
Closes https://jira.dhis2.org/browse/DHIS2-9866

Sorry for the large diff for what should have been a small change :sweat_smile:! The sent SMS section and the received SMS section have very similar behaviour but they were implemented slightly differently (the sent section uses the context API for refetching data while the received section uses `react-dom-router`) and so to ensure they both store filter params in the URL in the same way I had to extract the `Pagination` component and `useQueryParams` hook from the received SMS section.

### CODE REVIEW

File: src/views/delete_selected_button/DeleteSelectedButton.js [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
+}) => {
     const { addAlert } = useContext(AlertContext)
+    const mutation = useMemo(
+        () => ({
+            resource: mutationResource,
+            type: 'delete',
+            params: ({ ids }) => ({ ids }),
+        }),
+        [mutationResource]
+    )
```
Comment (ismay) [inline src/views/delete_selected_button/DeleteSelectedButton.js:21]: I would omit useMemo here until we really have a _need_ to optimize performance (and are measuring performance). Based on the recommendations here: https://kentcdodds.com/blog/usememo-and-usecallback

Comment (mediremi) [reply src/views/delete_selected_button/DeleteSelectedButton.js:21]: You're right that this wouldn't be useful for performance reasons, I only used `useMemo` because `useDataMutation` [complains if its first arg changes between renders](https://github.com/dhis2/app-runtime/blob/15b8344b829d88552f0dae337c97f83a902585d1/services/data/src/react/hooks/useStaticInput.ts#L16).

**Review by ismay — COMMENTED** [review]
> The sent SMS section and the received SMS section have very similar behaviour but they were implemented slightly differently (the sent section uses the context API for refetching data while the received section uses react-dom-router).

Yeah you're right. Reason for that is that this app was a bit of a rush job at the time and different people handled different sections. I worked on the section that doesn't use query params, and @HendrikThePendric worked on the one with the query params. We noticed the divergence but didn't resolve that before the release of the app.

One thing that I wasn't aware of is that we settled on using query params for both routes. I thought that was still undecided. Was that in the requirements for this app?

File: src/views/delete_selected_button/DeleteSelectedButton.js (hunk 1 above)
Comment (ismay) [reply src/views/delete_selected_button/DeleteSelectedButton.js:21]: Ah I see. Yeah there is a canonical way of doing this with the data mutation hook where it doesn't complain. Let me see if I can find an example.

**Comment by mediremi** [comment]
> One thing that I wasn't aware of is that we settled on using query params for both routes. I thought that was still undecided. Was that in the requirements for this app?

So right now the received SMS section uses the `useHistory` hook from `react-router-dom` to update query params, which then cause a refetch. Updating the query params in this way allows users to share links for filtered pages. On the other hand, the sent SMS section refetches using the context API but does not update the URL.

According to [DHIS2-9866](https://jira.dhis2.org/browse/DHIS2-9866), using query params for both routes is a requirement so that users can share links for filtered pages and so this PR updates the sent SMS section to use the same implementation of URL updating and refetching as the one used for the received SMS section.

File: src/views/delete_selected_button/DeleteSelectedButton.js (hunk 1 above)
Comment (ismay) [reply src/views/delete_selected_button/DeleteSelectedButton.js:21]: Sorry, that took a while, I'm doing too many things at the same time. So the docs don't mention this usecase explicitly. I took a look at the source and the only parts that can be dynamic (as far as I can tell) are id, data and params. So in this case since you want to use two different endpoints, depending on the type you want to delete, I'd use two queries.

The resource part of the query/mutation being static is by design, so I think that that also fits with Austin's intented usage of this hook.

**Comment by ismay** [comment]
> > One thing that I wasn't aware of is that we settled on using query params for both routes. I thought that was still undecided. Was that in the requirements for this app?
> 
> So right now the received SMS section uses the `useHistory` hook from `react-router-dom` to update query params, which then cause a refetch. Updating the query params in this way allows users to share links for filtered pages. On the other hand, the sent SMS section refetches using the context API but does not update the URL.
> 
> According to [DHIS2-9866](https://jira.dhis2.org/browse/DHIS2-9866), using query params for both routes is a requirement so that users can share links for filtered pages and so this PR updates the sent SMS section to use the same implementation of URL updating and refetching as the one used for the received SMS section.

Yeah I was commenting more on the requirement set in DHIS2-9866, which states that those changes should be made. Maybe I'm out of the loop, but at the time we made the app it was unclear to me whether this is something we wanted to do or not. So my question was whether that original question had been resolved without me knowing about it. It's not really something you can do anything about, what you've done fits perfectly with the request stated in the issue.

Since we already have this, and noone has commented on the changes during the meeting I'll assume the leads are ok with it and review accordingly.

**Comment by ismay** [comment]
A more general comment: some of the abstractions (like pagination) have been placed in views. Now if I understand the organisation of the app correctly, the generic components are all housed in `/src`. Maybe they'd be more appropriate there since they're now used across multiple routes.

That leads me to something else I noticed. The organisation in this app is a bit hard to follow. `./src` contains a lot of components, and the folders are camelcased. And then the folders in views are camelcased. I think we should create a follow up ticket to address these issues.

File: src/views/received_sms_list/SmsTable.js [hunk 2]
```diff
@@ -43,7 +42,11 @@ const SmsTable = ({ messages, pager, selectedIds, setSelectedIds }) => {
             <TableHead>
                 <TableRowHead>
                     <TableCellHead>
-                        <Checkbox checked={allSelected} onChange={toggleAll} />
+                        <Checkbox
+                            disabled={messages.length === 0}
+                            onChange={toggleAll}
+                            checked={allSelected}
+                        />
```
Comment (ismay) [inline src/views/received_sms_list/SmsTable.js:49]: This change is unrelated right? Not a huge deal if it is, but as a reviewer I'd prefer that in a separate commit for ease of reviewing (and finding the change later, if necessary).

**Comment by mediremi** [comment]
>  the generic components are all housed in /src. Maybe they'd be more appropriate there since they're now used across multiple routes.

:+1: 

> The organisation in this app is a bit hard to follow. ./src contains a lot of components, and the folders are camelcased. And then the folders in views are snakecased. That seems inconsistent to me. I think we should create a follow up ticket to address these issues.

Yep organising this app a bit more like e.g. [import-export-app](https://github.com/dhis2/import-export-app/tree/master/src) would help a lot.

File: src/views/received_sms_list/SmsTable.js (hunk 2 above)
Comment (mediremi) [reply src/views/received_sms_list/SmsTable.js:49]: That's right it was just part of trying to make components between both sections more similar. I should have moved these kind of changes to a different commit/PR for sure :sweat_smile:

File: i18n/en.pot [hunk 3]
```diff
@@ -136,6 +139,27 @@ msgstr ""
...
+msgstr ""
+
+msgid "Page"
+msgstr ""
+
+msgid "of {{pageCount}}"
+msgstr ""
+
+msgid "Next"
+msgstr ""
```
Comment (ismay) [inline i18n/en.pot:161]: We should create a follow up issue for the translations here that are part of a single sentence, but split up into separate translations. I don't think that this will lend itself well to translation to every language.

File: src/views/received_sms_list/ReceivedSmsList.js [hunk 4]
```diff
@@ -7,47 +7,49 @@ import {
...
-                filters.push(`originator:ilike:${phoneNumber}`)
-            }
-            if (status && status !== STATUS_ALL) {
-                filters.push(`smsstatus:eq:${status}`)
-            }
+    const filters = []
+    if (phoneNumber) {
+        filters.push(`originator:ilike:${phoneNumber}`)
+    }
+    if (status && status !== STATUS_ALL) {
```
Comment (ismay) [inline src/views/received_sms_list/ReceivedSmsList.js:38]: I see you've replaced the STATUS_ALL var with its value 'ALL' in other areas. I was wondering why you didn't do that here.

File: src/pagination/Pagination.js [hunk 5]
```diff
@@ -3,25 +3,36 @@ import { PropTypes } from '@dhis2/prop-types'
...
-    const { status, phoneNumber } = useQueryParams()
+    const queryParams = useQueryParams()
     const history = useHistory()
     const navigateToPage = newPage => {
         history.push({
             search: createSearchString({
-                status,
-                phoneNumber,
+                status: queryParams.status,
+                ...take(queryParams, extraSearchParams),
```
Comment (ismay) [inline src/pagination/Pagination.js:35]: If you spread the queryParams object here you should be able to just overwrite the params that need to be changed afterwards right? Or am I overlooking something?

File: src/views/received_sms_list/ReceivedSmsList.js [hunk 6]
```diff
@@ -7,47 +7,49 @@ import {
     RECEIVED_SMS_LIST_PATH,
     STATUS_ALL,
 } from './config'
-import { DeleteSelectedButton } from './DeleteSelectedButton'
+import DeleteSelectedButton from '../../delete_selected_button/DeleteSelectedButton'
```
Comment (ismay) [inline src/views/received_sms_list/ReceivedSmsList.js:10]: I'm all for default exports when only a single module is being exported from a file. But it doesn't seem to fit with the other components, judging from the imports here. I'd change all of them at once instead of introducing it just for a couple of files.

File: src/views/sent_sms_list/SentSmsList.module.css [hunk 7]
```diff
@@ -7,7 +7,3 @@
     align-items: flex-end;
     margin-bottom: var(--spacers-dp16);
 }
-
-.headerRight {
-    margin-left: auto;
-}
```
Comment (ismay) [inline src/views/sent_sms_list/SentSmsList.module.css:13]: It seems that the delete button is now left aligned because of this change. That introduces a bit of a discrepancy between the layout of both routes, as in the other route it is right aligned. I was wondering why this was changed.

File: src/views/sent_sms_list/SmsTable.js [hunk 8]
```diff
@@ -12,69 +12,82 @@ import {
...
+
+        setSelectedIds(Array.from(selectedIdSet))
+    }
+    const toggleAll = () => {
+        if (allSelected) {
+            setSelectedIds([])
+        } else {
+            setSelectedIds(messages.map(({ id }) => id))
+        }
+    }
```
Comment (ismay) [inline src/views/sent_sms_list/SmsTable.js:38]: It seems that these helper functions are now present for both the sent and the received sms tables. I think we could abstract those to a shared set of helpers for consistency, since they seem to fulfill the exact same role.

File: src/views/sent_sms_list/StatusFilter.js [hunk 9]
```diff
@@ -1,68 +1,48 @@
...
     return (
         <SingleSelectField
             label={i18n.t('Filter by status')}
             inputWidth="200px"
-            onChange={onChange}
+            onChange={handleStatusChange}
             selected={status}
         >
-            {filterOptions.map(({ label, value }) => (
+            {STATUS_FILTER_OPTIONS.map(({ label, value }) => (
```
Comment (ismay) [inline src/views/sent_sms_list/StatusFilter.js:41]: So this is why I was asking about everyone's preferences on uppercasing in the meeting. Since we don't have an established best practice there this seems fine to me.

**Review by ismay — COMMENTED** [review]
Thank you for the PR btw! Apologies for the number of comments. I hope that this is helpful, and let me know what your thoughts are on my remarks.

A couple of general things I noticed not related to any specific line:

- It seems like the filtering by status on the received route no longer works. Is this a regression, or was this already the case?
- I see we don't have a lot of unit tests. I think adding those is high priority. Ideally we'd at least cover the most important changes we made here with tests. Do you think that that would be attainable? Let me know what you think. It would increase my confidence in us not having broken anything.
- It seems like the phone filter wasn't added to the sent route, and that the layout differs slightly. Is this something we'll address in a follow up, or are we intending to keep these differences?

I think you did a good job btw, because this app is a little messy still from rushing it through release, so it's not terribly easy to work with yet. Let me know when I should rereview!

File: i18n/en.pot (hunk 3 above)
Comment (mediremi) [reply i18n/en.pot:161]: https://jira.dhis2.org/browse/DHIS2-10348

File: src/views/received_sms_list/ReceivedSmsList.js (hunk 4 above)
Comment (mediremi) [reply src/views/received_sms_list/ReceivedSmsList.js:38]: I replaced `STATUS_ALL` with `'ALL'` for the shared behaviour like `createSearchString` since `STATUS_ALL` was only defined in the `receved_sms_list` view. I was thinking of delaying more extensive refactoring to another PR but I might as well do it now.

File: src/pagination/Pagination.js (hunk 5 above)
Comment (mediremi) [reply src/pagination/Pagination.js:35]: You're right that this isn't needed - I thought I'd need this because `useQueryParams` returns `phoneNumber` regardless of whether it's actually used and I didn't want a `phoneNumber` query param to appear on the sent SMS list but it turns out that `createSearchString` filters out empty strings.

File: src/views/received_sms_list/ReceivedSmsList.js (hunk 6 above)
Comment (mediremi) [reply src/views/received_sms_list/ReceivedSmsList.js:10]: :100: I'll convert everything to a default export in a different PR and revert `DeleteSelectedButton` back to a single var export.

File: src/views/sent_sms_list/SentSmsList.module.css (hunk 7 above)
Comment (mediremi) [reply src/views/sent_sms_list/SentSmsList.module.css:13]: I was trying to make the CSS of both the sent and receive lists more similar and I forgot to add `flex-grow: 1` to the single select field on the sent SMS list.

File: src/views/sent_sms_list/SmsTable.js (hunk 8 above)
Comment (mediremi) [reply src/views/sent_sms_list/SmsTable.js:38]: Sounds good - I'll try to make a reusable `SmsTable` component that takes a child render function or just extract these into a group of selectors that can be imported.

File: src/views/sent_sms_list/StatusFilter.js (hunk 9 above)
Comment (mediremi) [reply src/views/sent_sms_list/StatusFilter.js:41]: I made this one uppercase since that's how it was defined in the receive SMS list but yeath we definitely need some kind of linter rule at some point.

**Comment by mediremi** [comment]
> It seems like the filtering by status on the received route no longer works. Is this a regression, or was this already the case?

It's unintuive but the current behaviour is that on the sent route refetching happens the moment that the status filter is changed, while on the receive route you have to click the 'filter' button to the right of the phone number input. I wasn't sure which behaviour to use for both routes since while having immediate status filtering is more intuitive, on the receive route it makes sense to not filter until the user submits since we don't want to fire off a network request everytime they input a phone number digit. 

Maybe it would make sense to make both status filters cause an immediate refetch and then rename the submit button on the receive route to 'filter by phone number' instead of just 'filter'? Or even use `onChange` instead of `onSubmit` on the receive route but debounce it?

>  I see we don't have a lot of unit tests. I think adding those is high priority. Ideally we'd at least cover the most important changes we made here with tests. Do you think that that would be attainable? Let me know what you think. It would increase my confidence in us not having broken anything.

I'd be happy to add cypress tests for the routes and unit tests for `createSearchString`.

> It seems like the phone filter wasn't added to the sent route, and that the layout differs slightly. Is this something we'll address in a follow up, or are we intending to keep these differences?

If the backend API supports filtering by recipient I'll add it in this PR and otherwise we can file a frontend and a backend issue on Jira to implement it.

File: i18n/en.pot (hunk 3 above)
Comment (ismay) [reply i18n/en.pot:161]: Thanks!

File: src/views/received_sms_list/ReceivedSmsList.js (hunk 4 above)
Comment (ismay) [reply src/views/received_sms_list/ReceivedSmsList.js:38]: Ah yes, that makes sense. All right, whichever you prefer. Leaving it for another PR if it's going to be too much for this one sounds fine to me, if you'd prefer that as well.

File: src/pagination/Pagination.js (hunk 5 above)
Comment (ismay) [reply src/pagination/Pagination.js:35]: I see. Assuming that we'll also have a phone number filter on the sent route we should then be able to simplify this area a bit.

File: src/views/received_sms_list/ReceivedSmsList.js (hunk 6 above)
Comment (ismay) [reply src/views/received_sms_list/ReceivedSmsList.js:10]: Thanks, sounds good!

File: src/views/sent_sms_list/SmsTable.js (hunk 8 above)
Comment (ismay) [reply src/views/sent_sms_list/SmsTable.js:38]: 👍

File: src/views/sent_sms_list/StatusFilter.js (hunk 9 above)
Comment (ismay) [reply src/views/sent_sms_list/StatusFilter.js]: I agree, going with the established convention in the app sounds like the best approach. (Since we as an organisation haven't decided on a recommendation yet.)

**Comment by ismay** [comment]
> > It seems like the filtering by status on the received route no longer works. Is this a regression, or was this already the case?
> 
> It's unintuive but the current behaviour is that on the sent route refetching happens the moment that the status filter is changed, while on the receive route you have to click the 'filter' button to the right of the phone number input. I wasn't sure which behaviour to use for both routes since while having immediate status filtering is more intuitive, on the receive route it makes sense to not filter until the user submits since we don't want to fire off a network request everytime they input a phone number digit.
> 
> Maybe it would make sense to make both status filters cause an immediate refetch and then rename the submit button on the receive route to 'filter by phone number' instead of just 'filter'? Or even use `onChange` instead of `onSubmit` on the receive route but debounce it?

Ah of course, you're right. I think we should definitely align the controls for both routes. And as to how we should change them exactly, I think it would be good to ask Joe. He'll be able to recommend what approach would make the most sense here.

> > I see we don't have a lot of unit tests. I think adding those is high priority. Ideally we'd at least cover the most important changes we made here with tests. Do you think that that would be attainable? Let me know what you think. It would increase my confidence in us not having broken anything.
> 
> I'd be happy to add cypress tests for the routes and unit tests for `createSearchString`.

That would be great!

> > It seems like the phone filter wasn't added to the sent route, and that the layout differs slightly. Is this something we'll address in a follow up, or are we intending to keep these differences?
> 
> If the backend API supports filtering by recipient I'll add it in this PR and otherwise we can file a frontend and a backend issue on Jira to implement it.

That sounds good, thanks 👍

**Comment by mediremi** [comment]
Jira issues for filtering sent SMS messages by recipeints: [DHIS2-10382](https://jira.dhis2.org/browse/DHIS2-10382) and [DHIS2-10383](https://jira.dhis2.org/browse/DHIS2-10383).

File: src/views/received_sms_list/Filter.js [hunk 10]
```diff
@@ -1,78 +1,133 @@
...
+    return (
+        <>
+            <InputField
+                label={i18n.t('Filter by phone number')}
+                inputWidth="250px"
+                onChange={handleChange}
+                value={value}
+                dataTest="phone-number-filter"
+            />
+            {waiting ? <CircularLoader small /> : null}
```
Comment (ismay) [inline src/views/received_sms_list/Filter.js:58]: You should be able to use the `loading` prop on the InputField for this.

File: src/views/received_sms_list/Filter.js [hunk 11]
```diff
@@ -1,78 +1,133 @@
...
+            onChange(event)
+            setWaiting(false)
+        }, 300),
+        [onChange]
+    )
+    const handleChange = ({ value }) => {
+        setValue(value)
+        setWaiting(true)
+        debouncedOnChange({ value })
+    }
```
Comment (ismay) [inline src/views/received_sms_list/Filter.js:47]: I've sent you (Medi) a message on slack with regards to our handling of state in query params, would be nice to get rid of the multiple sources of truth and simplify logic like this a bit.

File: src/views/received_sms_list/Filter.js (hunk 10 above)
Comment (ismay) [reply src/views/received_sms_list/Filter.js:58]: I assume that you did this because the spacing of the loading spinner on the InputField is off? We should create an issue for that.

**Review by ismay — APPROVED** [review]
Looks great, thanks for adding the tests! We should create follow up issues for the things that came up during the review (if they don't exist already):

- Organisation of the app should be improved, currently slightly inconsistent (https://github.com/dhis2/sms-configuration-app/pull/207#issuecomment-767486329)
- This could be a point of discussion, but moving the named exports to default exports if there's only one export. Best solution would probably be a proposal to add this rule to our shared linting rules: https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/prefer-default-export.md

File: src/views/received_sms_list/Filter.js (hunk 10 above)
Comment (mediremi) [reply src/views/received_sms_list/Filter.js:58]: I didn't realise that `InputField` could have a `loading` prop :sweat_smile: But yeah we should create an issue for the loading spinner spacing + maybe have it moved inside the input field so that it does not shrink

**Comment by mediremi** [comment]
> Organisation of the app should be improved, currently slightly inconsistent
https://jira.dhis2.org/browse/DHIS2-10347

File: src/views/received_sms_list/Filter.js (hunk 10 above)
Comment (ismay) [reply src/views/received_sms_list/Filter.js]: > maybe have it moved inside the input field so that it does not shrink

Actually, that was the old behaviour 😋. We ran into some issues there, which is why we have what we have now. But yeah, the spacing seems like a layout bug in the @dhis2/ui component.
